ql/src/java/org/apache/hadoop/hive/ql/Context.java

```diff
@@ -361,8 +361,10 @@ private DestClausePrefix getMergeDestClausePrefix(ASTNode curNode) {
     assert insert != null && insert.getType() == HiveParser.TOK_INSERT;
     ASTNode query = (ASTNode) insert.getParent();
     assert query != null && query.getType() == HiveParser.TOK_QUERY;
-
-    int tokFromIdx = query.getFirstChildWithType(HiveParser.TOK_FROM).getChildIndex();
+    ASTNode from = (ASTNode) query.getFirstChildWithType(HiveParser.TOK_FROM);
+    assert from != null : "Couldn't find a child of type FROM in the AST";
```

Comment:
Assertions should never be used for user-level errors. Despite the fact that they are disabled in production environments (and thus a user will never encounter) stack-traces and other information associated with the assertion/exception are for developers to help diagnose and fix a problem. Problems with the SQL syntax should be user-level errors with appropriate error codes and localization.

However, as I mentioned previously the `MERGE` statement addressed here is valid SQL so it shouldn't lead to failures or exceptions of any kind.